File: plover_vm/machine.py
"""Plover v0.1 virtual machine."""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path

from plover_vm.loader import load_hex, load_sram_program
from plover_vm.macro.engine import MacroEngine
from plover_vm.macro.fast import MacroFastPath
from plover_vm.memory.bus import MemoryBus
from plover_vm.micro.cw import lookup_cw
from plover_vm.micro.engine import MicroEngine
from plover_vm.trace import Tracer

EngineKind = str  # "micro" | "macro" | "fast"


@dataclass
class MachineState:
    pc: int = 0
    regs: list[int] = field(default_factory=lambda: [0, 0, 0, 0])
    map_mode: int = 0
    halted: bool = False
    phase: int = 0
    opcode: int = 0


class PloverMachine:
    def __init__(self, engine: EngineKind = "micro") -> None:
        self.bus = MemoryBus()
        self.micro = MicroEngine(self.bus)
        self.macro = MacroEngine(self.bus, self.micro)
        self.fast = MacroFastPath(self.bus)
        self.engine = engine
        self.tracer = Tracer()
        self._steps = 0

    def load_nor(self, path: Path | str, offset: int = 0) -> None:
        self.bus.nor.load_hex(path, offset)

    def load_cw(self, path: Path | str) -> None:
        import sys

        root = Path(__file__).resolve().parents[1]
        if str(root) not in sys.path:
            sys.path.insert(0, str(root))
        from tools.pack_control_store import CW_FLASH_BASE, build_all

        data, _ = load_hex(path, 0)
        if data:
            self.bus.nor.patch_cw_region(list(data), CW_FLASH_BASE)
        else:
            self.bus.nor.patch_cw_region(build_all(), CW_FLASH_BASE)

    def load_ram_program(self, path: Path | str, base: int = 0) -> None:
        prog = load_sram_program(path)
        self.bus.ram.load(prog, base)

    def load_ram_bytes(self, data: bytes | bytearray, base: int = 0) -> None:
        self.bus.ram.load(data, base)

    def set_map_mode(self, mode: int) -> None:
        self.bus.map_mode = mode & 1

    def reset(self, map_mode: int | None = None) -> None:
        if map_mode is not None:
            self.bus.map_mode = map_mode & 1
        fa = 0xFFFC
        lo = self.bus.read_cpu(fa)
        hi = self.bus.read_cpu((fa + 1) & 0xFFFF)
        entry = lo | (hi << 8)

        self.macro.pc = entry & 0xFFFF
        self.macro.halted = False
        self.macro._fetch_pending = True
        self.macro._ret_stack = []
        self.micro.state = self.micro.state.__class__()
        self.fast.pc = entry & 0xFFFF
        self.fast.halted = False
        self.fast.regs = [0, 0, 0, 0]
        self.fast._ret_stack = []
        self._steps = 0

    def step_once(self) -> None:
        if self.engine == "fast":
            self.fast.step()
            self._record_trace_fast()
            return

        self.macro.step()
        self._record_trace_micro()

    def _record_trace_micro(self) -> None:
        st = self.micro.state
        op = self.macro.opcode
        ph = max(0, st.phase - 1)
        cw = lookup_cw(self.bus.nor.read_cw, op, ph).raw if op else 0
        self.tracer.record(
            self.macro.pc,
            st.phase,
            op,
            cw,
            st.regs,
            self.macro.halted,
        )
        self._steps += 1

    def _record_trace_fast(self) -> None:
        self.tracer.record(
            self.fast.pc,
            0,
            0,
            0,
            self.fast.regs,
            self.fast.halted,
        )
        self._steps += 1

    def run(self, max_steps: int = 10_000) -> MachineState:
        for _ in range(max_steps):
            if self.halted:
                break
            self.step_once()
        return self.snapshot()

    @property
    def halted(self) -> bool:
        if self.engine == "fast":
            return self.fast.halted
        return self.macro.halted

    def snapshot(self) -> MachineState:
        if self.engine == "fast":
            return MachineState(
                pc=self.fast.pc,
                regs=list(self.fast.regs),
                map_mode=self.bus.map_mode,
                halted=self.fast.halted,
            )
        st = self.micro.state
        return MachineState(
            pc=self.macro.pc,
            regs=list(st.regs),
            map_mode=self.bus.map_mode,
            halted=self.macro.halted,
            phase=st.phase,
            opcode=self.macro.opcode,
        )

    @classmethod
    def from_config(
        cls,
        *,
        nor: Path | None = None,
        cw: Path | None = None,
        ram_prog: Path | None = None,
        map_mode: int = 0,
        engine: EngineKind = "micro",
    ) -> PloverMachine:
        m = cls(engine=engine)
        root = Path(__file__).resolve().parents[1]
        if nor:
            m.load_nor(nor)
        else:
            boot = root / "hw" / "fixtures" / "boot" / "boot_rom.hex"
            if boot.is_file():
                m.load_nor(boot, 0)
            vec = root / "hw" / "fixtures" / "boot" / "boot_vector.hex"
            if vec.is_file():
                m.load_nor(vec, 0xFFFC)
        if cw:
            m.load_cw(cw)
        else:
            cw_path = root / "hw" / "fixtures" / "control" / "cw.hex"
            if cw_path.is_file():
                m.load_cw(cw_path)
            else:
                m.load_cw(Path("/dev/null"))
        if ram_prog:
            m.load_ram_program(ram_prog, 0x0800 if map_mode else 0)
        m.bus.map_mode = map_mode
        return m

File: crates/plover_core/src/machine.rs
use crate::fast::MacroFastPath;
use crate::macro_eng::MacroEngine;
use crate::micro::{lookup_cw, MicroEngine};
use crate::trace::Tracer;
use plover_mmu::hex::load_hex;
use plover_mmu::{hex::load_sram_program, CW_FLASH_BASE, MemoryBus};
use std::path::Path;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
pub enum EngineKind {
    #[default]
    Micro,
    Macro,
    Fast,
}

impl EngineKind {
    pub fn parse(s: &str) -> Self {
        match s.to_ascii_lowercase().as_str() {
            "macro" => Self::Macro,
            "fast" => Self::Fast,
            _ => Self::Micro,
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct MachineSnapshot {
    pub pc: u16,
    pub regs: [u8; 4],
    pub halted: bool,
    pub map_mode: u8,
    pub flag_z: bool,
    pub flag_c: bool,
    pub phase: u8,
    pub opcode: u8,
}

pub struct PloverMachine {
    pub bus: MemoryBus,
    pub micro: MicroEngine,
    pub macro_eng: MacroEngine,
    pub fast: MacroFastPath,
    pub engine: EngineKind,
    pub tracer: Tracer,
}

impl PloverMachine {
    pub fn new() -> Self {
        Self::with_engine(EngineKind::Micro)
    }

    pub fn with_engine(engine: EngineKind) -> Self {
        let bus = MemoryBus::default();
        let micro = MicroEngine::new(bus.clone());
        let macro_eng = MacroEngine::new(bus.clone());
        let fast = MacroFastPath::new(bus.clone());
        Self {
            bus,
            micro,
            macro_eng,
            fast,
            engine,
            tracer: Tracer::default(),
        }
    }

    pub fn set_engine(&mut self, engine: EngineKind) {
        self.engine = engine;
    }

    pub fn load_nor(&mut self, path: &Path, offset: usize) {
        self.bus.nor.load_hex_file(path, offset);
        self.sync_all_buses();
    }

    pub fn load_cw(&mut self, path: &Path) {
        let data = load_hex(path, 0);
        if data.is_empty() {
            let root = Path::new(env!("CARGO_MANIFEST_DIR")).join("../..");
            let fallback = root.join("hw/fixtures/control/cw.hex");
            let data = load_hex(&fallback, 0);
            self.bus.nor.patch_cw_region(&data, CW_FLASH_BASE);
        } else {
            self.bus.nor.patch_cw_region(&data, CW_FLASH_BASE);
        }
        self.sync_all_buses();
    }

    pub fn load_default_boot_fixtures(&mut self, root: &Path) {
        let boot = root.join("hw/fixtures/boot");
        let rom = boot.join("boot_rom.hex");
        if rom.is_file() {
            self.load_nor(&rom, 0);
        }
        let vec = boot.join("boot_vector.hex");
        if vec.is_file() {
            self.load_nor(&vec, 0xFFFC);
        }
        let cw = root.join("hw/fixtures/control/cw.hex");
        if cw.is_file() {
            self.load_cw(&cw);
        }
    }

    pub fn set_map_mode(&mut self, mode: u8) {
        self.bus.map_mode = mode & 1;
        self.sync_all_buses();
    }

    pub fn set_pc(&mut self, pc: u16) {
        self.fast.pc = pc;
        self.macro_eng.pc = pc;
    }

    pub fn set_regs(&mut self, regs: [u8; 4]) {
        self.fast.regs = regs;
        self.micro.state.regs = regs;
        self.macro_eng.micro.state.regs = regs;
    }

    pub fn regs(&self) -> [u8; 4] {
        match self.engine {
            EngineKind::Fast => self.fast.regs,
            EngineKind::Micro | EngineKind::Macro => self.macro_eng.micro.state.regs,
        }
    }

    pub fn halted(&self) -> bool {
        match self.engine {
            EngineKind::Fast => self.fast.halted,
            EngineKind::Micro | EngineKind::Macro => self.macro_eng.halted,
        }
    }

    pub fn pc(&self) -> u16 {
        match self.engine {
            EngineKind::Fast => self.fast.pc,
            EngineKind::Micro | EngineKind::Macro => self.macro_eng.pc,
        }
    }

    pub fn load_ram(&mut self, data: &[u8], base: u16) {
        self.bus.ram.load(data, base);
        self.sync_all_buses();
    }

    pub fn load_ram_bytes(&mut self, data: &[u8], base: u16) {
        self.load_ram(data, base);
    }

    pub fn reset_exec_state(&mut self) {
        self.fast.regs = [0, 0, 0, 0];
        self.fast.halted = false;
        self.fast.pc = 0;
        self.fast.flag_z = false;
        self.fast.flag_c = false;
        self.micro.state = Default::default();
        self.macro_eng.halted = false;
        self.macro_eng.pc = 0;
        self.macro_eng.fetch_pending = true;
        self.macro_eng.ret_stack.clear();
    }

    pub fn reset(&mut self, map_mode: Option<u8>) {
        if let Some(m) = map_mode {
            self.set_map_mode(m);
        }
        let lo = self.bus.read_cpu(0xFFFC);
        let hi = self.bus.read_cpu(0xFFFD);
        let entry = u16::from(lo) | (u16::from(hi) << 8);
        self.macro_eng.pc = entry;
        self.macro_eng.halted = false;
        self.macro_eng.fetch_pending = true;
        self.macro_eng.ret_stack.clear();
        self.micro.state = Default::default();
        self.fast.pc = entry;
        self.fast.halted = false;
        self.fast.regs = [0, 0, 0, 0];
        self.fast.ret_stack.clear();
        self.sync_all_buses();
    }

    pub fn snapshot(&self) -> MachineSnapshot {
        match self.engine {
            EngineKind::Fast => MachineSnapshot {
                pc: self.fast.pc,
                regs: self.fast.regs,
                map_mode: self.bus.map_mode,
                halted: self.fast.halted,
                flag_z: self.fast.flag_z,
                flag_c: self.fast.flag_c,
                phase: 0,
                opcode: 0,
            },
            EngineKind::Micro | EngineKind::Macro => {
                let st = &self.macro_eng.micro.state;
                MachineSnapshot {
                    pc: self.macro_eng.pc,
                    regs: st.regs,
                    map_mode: self.bus.map_mode,
                    halted: self.macro_eng.halted,
                    flag_z: st.flag_z,
                    flag_c: st.flag_c,
                    phase: st.phase,
                    opcode: self.macro_eng.opcode(),
                }
            }
        }
    }

    pub fn ram_nonzero(&self) -> usize {
        self.bus.ram.nonzero_count()
    }

    pub fn load_ram_program(&mut self, path: &Path, base: u16) {
        let data = load_sram_program(path);
        self.load_ram(&data, base);
    }

    pub fn sync_buses(&mut self) {
        self.sync_all_buses();
    }

    fn sync_all_buses(&mut self) {
        self.fast.bus = self.bus.clone();
        self.micro.bus = self.bus.clone();
        self.macro_eng.bus = self.bus.clone();
        self.macro_eng.micro.bus = self.bus.clone();
    }

    fn pull_bus(&mut self) {
        match self.engine {
            EngineKind::Fast => self.bus = self.fast.bus.clone(),
            EngineKind::Micro | EngineKind::Macro => self.bus = self.macro_eng.bus.clone(),
        }
        self.micro.bus = self.bus.clone();
        self.fast.bus = self.bus.clone();
    }

    pub fn step_once(&mut self) {
        match self.engine {
            EngineKind::Fast => {
                self.sync_all_buses();
                self.fast.step();
                self.pull_bus();
                self.tracer.record(
                    self.fast.pc,
                    0,
                    0,
                    0,
                    self.fast.regs,
                    self.fast.halted,
                );
            }
            EngineKind::Micro | EngineKind::Macro => {
                self.sync_all_buses();
                self.macro_eng.step();
                self.pull_bus();
                self.micro.state = self.macro_eng.micro.state.clone();
                let st = &self.macro_eng.micro.state;
                let op = self.macro_eng.opcode();
                let ph = st.phase.saturating_sub(1);
                let cw = if op != 0 {
                    lookup_cw(|idx| self.bus.nor.read_cw(idx), op, ph).raw
                } else {
                    0
                };
                self.tracer.record(
                    self.macro_eng.pc,
                    st.phase,
                    op,
                    cw,
                    st.regs,
                    self.macro_eng.halted,
                );
            }
        }
    }

    pub fn run(&mut self, max_steps: usize) {
        for _ in 0..max_steps {
            if self.halted() {
                break;
            }
            self.step_once();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    fn repo_root() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../..")
    }

    fn setup_micro(root: &Path) -> PloverMachine {
        let mut m = PloverMachine::with_engine(EngineKind::Micro);
        m.load_cw(&root.join("hw/fixtures/control/cw.hex"));
        m
    }

    #[test]
    fn add_imm_fast() {
        let root = repo_root();
        let mut m = PloverMachine::with_engine(EngineKind::Fast);
        m.set_map_mode(1);
        m.set_regs([0x12, 0, 0, 0]);
        m.set_pc(0);
        m.load_ram_program(&root.join("hw/fixtures/sram/add_imm.sram.hex"), 0);
        m.run(500);
        assert!(m.halted());
        assert_eq!(m.regs(), [0x12, 0x34, 0x46, 0]);
    }

    #[test]
    fn add_imm_micro() {
        let root = repo_root();
        let mut m = setup_micro(&root);
        m.set_map_mode(1);
        m.set_regs([0x12, 0, 0, 0]);
        m.set_pc(0);
        m.load_ram_program(&root.join("hw/fixtures/sram/add_imm.sram.hex"), 0);
        m.run(500);
        assert!(m.halted());
        assert_eq!(m.regs(), [0x12, 0x34, 0x46, 0]);
    }

    #[test]
    fn call_ret_parity() {
        let root = repo_root();
        let mut m = PloverMachine::with_engine(EngineKind::Fast);
        m.set_map_mode(1);
        m.set_pc(0);
        m.load_ram_program(&root.join("hw/fixtures/sram/call_ret.sram.hex"), 0);
        m.run(200);
        assert!(m.halted());
        assert_eq!(m.regs()[0], 11);
    }

    #[test]
    fn micro_matches_fast_mov() {
        let root = repo_root();
        let prog = root.join("hw/fixtures/sram/add_imm.sram.hex");
        let mut fast = PloverMachine::with_engine(EngineKind::Fast);
        fast.set_map_mode(1);
        fast.set_regs([0x12, 0, 0, 0]);
        fast.set_pc(0);
        fast.load_ram_program(&prog, 0);
        fast.run(500);

        let mut micro = setup_micro(&root);
        micro.set_map_mode(1);
        micro.set_regs([0x12, 0, 0, 0]);
        micro.set_pc(0);
        micro.load_ram_program(&prog, 0);
        micro.run(500);

        assert_eq!(fast.regs(), micro.regs());
        assert_eq!(fast.halted(), micro.halted());
    }
}
